src/helpers/userHelper.js

```diff
@@ -965,133 +965,93 @@ const userHelper = function () {
-                  addBadge(personId, mongoose.Types.ObjectId(bdge._id));
-                  removePrevHrBadge(
-                    personId,
-                    user,
-                    badgeCollection,
-                    bdge.hrs,
-                    bdge.weeks,
-                  );
-                } else if (badgeOfType && badgeOfType.totalHrs === bdge.hrs) {
-                  increaseBadgeCount(
-                    personId,
-                    mongoose.Types.ObjectId(badgeOfType._id),
-                  );
-                  removePrevHrBadge(
-                    personId,
-                    user,
-                    badgeCollection,
-                    bdge.hrs,
-                    bdge.weeks,
-                  );
-                }
-                return false;
-              }
-            }
-            return true;
-          });
-        });
-      });
-  };
+   // Here it is verifying the maximun  hour in the array of badges for the first condition
+   // if the user worked 45 hours in a week just make sense assign badges where the total hours are lower then or equal 45 hours
+   const maxHourBadge = Math.max(...arrayTotalHrs);
+   // now here will assign the especific badge
+   arrayBadges.map((badge) => {
+     if (badge.totalHrs === maxHourBadge) {
+       let badgeId;
+       for (let i = 0; i < badgesOfType.length; i++) {
+         if (badgesOfType[i]._id.toString() === badge._id.toString()) {
+             badgeId = badgesOfType[i]._id;
+         }
+       }
+       if (badgeId) {
+         increaseBadgeCount(personId, mongoose.Types.ObjectId(badgeId));
+         return false;
+       }
+         addBadge(personId, mongoose.Types.ObjectId(badge._id));
+     }
+   });
+   // check for week greater than 1
+   // check how many weeks the user worked
+   // based on the length of user.savedTangibleHrs array
+   const workedWeeks = user.savedTangibleHrs.length;
+   const greatStreakBadges = await badge.find({ type: 'X Hours for X Week Streak', weeks: { $lte: workedWeeks, $gt: 1 } }).sort({ totalHrs: -1 });
+   // for the badge be assigned have to fulfill the requirement  totalHrs < user.lastWeekTangibleHrs;
+
+   const badgesThatMatch = [];
+   // this array is storing all the badges that fullfil the totalHrs < lastWeekTangibleHrs
+   greatStreakBadges.forEach((item) => {
+     if (item.totalHrs <= user.lastWeekTangibleHrs) {
+       badgesThatMatch.push(item);
+     }
+   });
+   const matchBadge = [];
+   const weeksArray = [];
+
+   badgeCollection.filter(bdg => bdg.badge?.type === 'X Hours for X Week Streak').map((badge) => {
+     for (let i = 0; i < badgesThatMatch.length; i++) {
+       if (badge.badge?.totalHrs === badgesThatMatch[i].totalHrs) {
+         matchBadge.push(badgesThatMatch[i]);
+         weeksArray.push(badgesThatMatch[i].weeks);
+       }
+     }
+   });
+   // calculate the max week number to assign the badge;
+   const maxWeek = Math.max(...weeksArray);
```

Comment:
Throwing in a conditional or ternary here to handle an edge case where weeksArray is empty will prevent infinity values being generated for this variable.

Reply:
done